File: strategy/php/Strategy.php
<?php

// Strategy interface
interface PaymentStrategy {
    public function pay(float $amount): string;
}

// Concrete Strategies
class CreditCardPayment implements PaymentStrategy {
    private $cardNumber;

    public function __construct(string $cardNumber) {
        $this->cardNumber = $cardNumber;
    }

    public function pay(float $amount): string {
        $lastFour = substr($this->cardNumber, -4);
        return sprintf("Paid $%.2f using Credit Card ending in %s", $amount, $lastFour);
    }
}

class PayPalPayment implements PaymentStrategy {
    private $email;

    public function __construct(string $email) {
        $this->email = $email;
    }

    public function pay(float $amount): string {
        return sprintf("Paid $%.2f using PayPal account %s", $amount, $this->email);
    }
}

class BitcoinPayment implements PaymentStrategy {
    private $walletAddress;

    public function __construct(string $walletAddress) {
        $this->walletAddress = $walletAddress;
    }

    public function pay(float $amount): string {
        $shortAddress = substr($this->walletAddress, 0, 8) . "...";
        return sprintf("Paid $%.2f using Bitcoin wallet %s", $amount, $shortAddress);
    }
}

// Context
class ShoppingCart {
    private $amount;
    private $paymentStrategy;

    public function __construct(float $amount) {
        $this->amount = $amount;
    }

    public function setPaymentStrategy(PaymentStrategy $strategy): void {
        $this->paymentStrategy = $strategy;
    }

    public function checkout(): string {
        return $this->paymentStrategy->pay($this->amount);
    }
}

// Usage
$cart = new ShoppingCart(100.0);

$cart->setPaymentStrategy(new CreditCardPayment("1234567890123456"));
echo $cart->checkout() . "\n";

$cart->setPaymentStrategy(new PayPalPayment("user@example.com"));
echo $cart->checkout() . "\n";

$cart->setPaymentStrategy(new BitcoinPayment("1A2B3C4D5E6F7G8H"));
echo $cart->checkout() . "\n";

File: patterns/strategy/python/Strategy.py
from abc import ABC, abstractmethod

# Strategy interface
class PaymentStrategy(ABC):
    @abstractmethod
    def pay(self, amount: float) -> str:
        pass

# Concrete Strategies
class CreditCardPayment(PaymentStrategy):
    def __init__(self, card_number: str):
        self._card_number = card_number

    def pay(self, amount: float) -> str:
        return f"Paid ${amount:.2f} using Credit Card ending in {self._card_number[-4:]}"

class PayPalPayment(PaymentStrategy):
    def __init__(self, email: str):
        self._email = email

    def pay(self, amount: float) -> str:
        return f"Paid ${amount:.2f} using PayPal account {self._email}"

class BitcoinPayment(PaymentStrategy):
    def __init__(self, wallet_address: str):
        self._wallet_address = wallet_address

    def pay(self, amount: float) -> str:
        return f"Paid ${amount:.2f} using Bitcoin wallet {self._wallet_address[:8]}..."

# Context
class ShoppingCart:
    def __init__(self, amount: float):
        self._amount = amount
        self._payment_strategy: PaymentStrategy = None

    def set_payment_strategy(self, strategy: PaymentStrategy):
        self._payment_strategy = strategy

    def checkout(self) -> str:
        return self._payment_strategy.pay(self._amount)


if __name__ == "__main__":
    cart = ShoppingCart(100.0)

    cart.set_payment_strategy(CreditCardPayment("1234567890123456"))
    print(cart.checkout())

    cart.set_payment_strategy(PayPalPayment("user@example.com"))
    print(cart.checkout())

    cart.set_payment_strategy(BitcoinPayment("1A2B3C4D5E6F7G8H"))
    print(cart.checkout())
